storage/s3/src/main/java/io/aiven/kafka/tieredstorage/storage/s3/MetricCollector.java

```diff
@@ -55,41 +58,67 @@ class MetricCollector implements MetricPublisher {
+        latencyMetrics.put("CreateMultipartUpload", createLatencySensor("create-multipart-upload-time"));
+        requestMetrics.put("CompleteMultipartUpload", createRequestsSensor("complete-multipart-upload-requests"));
+        latencyMetrics.put("CompleteMultipartUpload", createLatencySensor("complete-multipart-upload-time"));
+        requestMetrics.put("PutObject", createRequestsSensor("put-object-requests"));
+        latencyMetrics.put("PutObject", createLatencySensor("put-object-time"));
+        requestMetrics.put("DeleteObject", createRequestsSensor("delete-object-requests"));
+        latencyMetrics.put("DeleteObject", createLatencySensor("delete-object-time"));
+        requestMetrics.put("AbortMultipartUpload", createRequestsSensor("abort-multipart-upload-requests"));
+        latencyMetrics.put("AbortMultipartUpload", createLatencySensor("abort-multipart-upload-time"));
+
+        errorMetrics.put(THROTTLING.toString(), createRequestsSensor("throttling-errors"));
+        errorMetrics.put(SERVER_ERROR.toString(), createRequestsSensor("server-errors"));
+        errorMetrics.put(CONFIGURED_TIMEOUT.toString(), createRequestsSensor("configured-timeout-errors"));
+        errorMetrics.put(IO.toString(), createRequestsSensor("io-errors"));
+        errorMetrics.put(OTHER.toString(), createRequestsSensor("other-errors"));
     }
 
-    private Sensor createSensor(final String name) {
+    private Sensor createRequestsSensor(final String name) {
         final Sensor sensor = metrics.sensor(name);
         sensor.add(metrics.metricName(name + "-rate", METRIC_GROUP), new Rate());
         sensor.add(metrics.metricName(name + "-total", METRIC_GROUP), new CumulativeCount());
         return sensor;
     }
 
+    private Sensor createLatencySensor(final String name) {
+        final Sensor sensor = metrics.sensor(name);
+        sensor.add(metrics.metricName(name + "-max", METRIC_GROUP), new Max());
+        sensor.add(metrics.metricName(name + "-avg", METRIC_GROUP), new Avg());
+        return sensor;
+    }
+
     @Override
     public void publish(final MetricCollection metricCollection) {
         final List<String> metricValues = metricCollection.metricValues(CoreMetric.OPERATION_NAME);
-        for (final String metricValue : metricValues) {
-            final var sensor = requestMetrics.get(metricValue);
-            if (sensor != null) {
-                sensor.record();
+        // metrics are reported per request, so 1 value can be assumed.
+        if (metricValues.size() == 1) {
+            final var metricValue = metricValues.get(0);
+            final var requests = requestMetrics.get(metricValue);
+            if (requests != null) {
+                requests.record();
+            }
+
+            final var durations = metricCollection.metricValues(CoreMetric.API_CALL_DURATION);
+            if (durations.size() == 1) {
```

Comment:
Does this need to be inside the previous if? And same question regarding number of values.

Reply:
Yeah, it's to get the right sensor, as the metric for latency only contains the duration. As the metric collection is correlated, then I'm using this value to update metrics properly.